### ISSUE TITLE
[Question] How could i insert my on php code and show it within a box in lansuite?

### ISSUE THREAD

**Issue by chimcen** [lansuite/lansuite#184]
## Context
I've a short question - i want to insert my php-code to load json-data from my discord server and show it in a box on my lansuite website... how could this be done?

Just for info - my php-code with the json-data fetch is already running on another website of mine (e107 cms) and is working...

**Comment by M4LuZ** [comment]
(German) documentation about how to generate box content can be found here:
http://lansuite.orgapage.de/index.php?mod=wiki&action=show&name=Box%20erstellen
Basically: 
Drop the code into a PHP file, either create a new module or drop it into an existing one and then configure the boxes shown (and what is to be called to generate them) via the Box-Manager.
The module for that may need to be enabled first.
If done you can use it to specify the module and file to be executed and where and for whom to display it.
And that should get you going.
Disclaimer: Never done it myself, so to be taken with a grain of salt.

**Comment by chimcen** [comment]
it works, but not as exprected...

I've created a new file "discord.php" wich contains my code and uploaded this to my server into stats module "boxes" folder.
After then i've created a new box inside boxmanager.
My Discord box will be displayed, but gets contents from another box... my discord server viewer will be displayed in the upper left corner of the page, instead inside the new created box...

![image](https://user-images.githubusercontent.com/22658487/38301282-66d0d37e-37ff-11e8-8133-6fddcdb54fa8.png)
![image](https://user-images.githubusercontent.com/22658487/38301359-959eaf64-37ff-11e8-92c5-a6434474613e.png)

**Comment by M4LuZ** [comment]
Can you please share the content of `discord.php` with us?
Remember to remove credentials first.

**Comment by chimcen** [comment]
`<?php
/**
 * Show discordbox
 *
 * @author CCG*Centurio
 * @version $Id: discord.php 1673 2018-04-04 08:13:47Z CCG*Centurio $
 */


	$json = file_get_contents('https://discordapp.com/api/guilds/277788790565634048/widget.json');
	$discord = json_decode($json, true);
	echo "<li class='discord_server_name'>{$discord['name']}<br>";
	
	$discord1 = json_decode(file_get_contents('https://discordapp.com/api/servers/277788790565634048/widget.json'));


// -------------------------------- MEMBERS ---------------------------------------- //	

	
	// echo '<pre>' . print_r($discord, true) . '</pre>';
	
if (count($discord['members']) > 0) {
	echo '<span class="online_users badge green">' . count($discord['members']) . '</span>';
}
else {
	echo '<span class="online_users badge red">' . count($discord['members']) . '</span>';
} 
echo '<ul class="online_sidebar">';
	foreach($discord['members'] as $member):
		if (array_key_exists('nick', $member)) {
			echo '<li><img src="'. $member['avatar_url'] .'" class="'. $member['status'] .' discord_avatar">' . $member['nick'] . '</li>';
		}
		else {
			echo '<li><img src="'. $member['avatar_url'] .'" class="'. $member['status'] .' discord_avatar">' . $member['username'] . '</li>';
		}
	endforeach;
echo '</ul>';


// -------------------------------- CHANNELS ---------------------------------------- //


  if ($discord1->channels) {
	usort($discord1->channels, function($a, $b) {
	  return ($a->position > $b->position) ? 1 : -1;
	});
	echo '<ul class="online_sidebar_channel">';
	foreach ($discord1->members as $member) {
		if (array_key_exists('nick', $member) && !empty($member->channel_id)) {
			$channel_members[$member->channel_id][] = $member->nick;
		}
		elseif (!empty($member->channel_id)) {
			$channel_members[$member->channel_id][] = $member->username;
		}
	}
	foreach ($discord1->channels as $channel) {
	  echo "<li class='channel'>{$channel->name}";
	  if (!empty($channel_members[$channel->id])) {
			echo '<ul>';
			foreach ($channel_members[$channel->id] as $username) {
			  echo "<li class='channel_member'>$username</li>";
			}
			echo '</ul>';
	  }
	  echo "</li>";
	}
	echo "<input class=\"btn-join\" type=button onClick=\"parent.open('https://discordapp.com/invite/qScqjph')\" value='Join'>";
	echo '</ul>';
  }
  
  
?>
`

**Comment by M4LuZ** [comment]
Solution for that is simple: Don't ignore the following bit of the answer ;)
> (German) documentation about how to generate box content can be found here:
http://lansuite.orgapage.de/index.php?mod=wiki&action=show&name=Box%20erstellen

That describes the functions to be used so that content is inserted at the right position within the template.
If you bypass the template engine by calling `echo`directly, then your content is shown before the main page structure is defined.

**Comment by chimcen** [comment]
Ok... can't open the link - gitting alwas "HTTP ERROR 502" - on every single divice i've tried... :(

Can you help me by changing my code to fit the template engine?

**Comment by M4LuZ** [comment]
I think we can rework this into a general module if you are happy with contributing your code for that

**Comment by chimcen** [comment]
Shure, why not - improvment for us all...
Do you use Gitter? There we can talk in detail about that topic - ...and in german...  ;)

**Comment by chimcen** [comment]
Any news?

**Comment by chimcen** [comment]
Awesome :) :+1:

**Comment by M4LuZ** [comment]
@chimcen Got you code up and running now.
Try pulling the code from the branch (don't forget to pull composer dependencies!) and give it a try.
* Activate Discord as a module
* Go to Configuration from the Module overview 
* Add server and Guild ID there (both seem to return the same JSON array, may be merged in the future...)
*  Enable Discord under boxes
* A discord box should appear on the right side with some content

Feedback (and further development) would be much appreciated

**Comment by chimcen** [comment]
sry, was a little bit busy at last...
btw. does the latest git of lansuite also run on php5.6 or is php7.x necessary?

**Comment by M4LuZ** [comment]
> btw. does the latest git of lansuite also run on php5.6 or is php7.x necessary?

From [The main page](https://github.com/lansuite/lansuite#requirements)
``` 
Requirements

        = PHP 7 (with mysqli, snmp and gd extensions)

        = MySQL 5.6.3
```

Yeah, I think it does ;)
But you should be able to get it to run with PHP 5.6 if you follow this:
> If you want to test it with LS 4.2 you would need to add an require_once 'modules/discord/Classes/Discord.php';in boxes/discord.php at the top. Not sure if Namespacing causes an issue.
If it complains, remove it both from Classes/Discord.php and from the instantiation in boxes/Discord.php

**Comment by andygrunwald** [comment]
@chimcen One additional information:
PHP 5.6 is not actively supported anymore.  Right now it only gets Security Updates.
The PHP Core team highly suggests updating to >= PHP 7.0.
See https://secure.php.net/supported-versions.php

**Comment by chimcen** [comment]
yes, i know - but actualy i've a running lansuite website running on php5.6 and lansuite 4.2 - i fear the upgrading to latest lansuite version...

**Comment by M4LuZ** [comment]
> i fear the upgrading to latest lansuite version...

At some point you should rather fear about not having updated given what was fixed in the meantime in both PHP and LS.
We'll provide a clear update guide for the next release and in general there is not a lot that can cause you an issue if you did your backup-homework ;)

**Comment by chimcen** [comment]
I'm waiting for the update guide before updating ;)

back to topic...

i used the new branched lansuite version with the discord plugin (only copied changed files to main branch on my server) and configured it with my server/guild id.

Box is configured in boxmanager and it appears on my site.

Issues:

1) lansuite calls a different id as stored in plugin config and shows a failure
(my id is 277788790565634048 ) :

   `Fehler: 
PHP Warning: file_get_contents(https://discordapp.com/api/servers/2147483647/widget.json): failed to open stream: HTTP request failed! HTTP/1.1 404 NOT FOUND in /volume1/web/lansuite/modules/discord/Classes/Discord.php on line 67
PHP Warning: file_get_contents(https://discordapp.com/api/guilds/2147483647/widget.json): failed to open stream: HTTP request failed! HTTP/1.1 404 NOT FOUND in /volume1/web/lansuite/modules/discord/Classes/Discord.php on line 61
PHP Warning: Invalid argument supplied for foreach() in /volume1/web/lansuite/modules/discord/Classes/Discord.php on line 88`

If i enter my server/guild id manually in Clases/Discord.php the plugin works


2) Server channels aren't displayed...
The plugin installed (LS 4.2):
![image](https://user-images.githubusercontent.com/22658487/40116977-c27a2dc4-5915-11e8-89d5-04c849ba00f1.png)


And how it looks like with channels and css themed (at another cms):
![image](https://user-images.githubusercontent.com/22658487/40117025-f4294b98-5915-11e8-9a24-3326d7721611.png)


3) theming problems (discord user avatar is shown at max resolution) - actually i checkout some css stuff for it...	


If needed you can write me a pm at gitter -> https://gitter.im/chimcen
(german is also possible)

**Comment by M4LuZ** [comment]
Well, I leave no.2 and 3 to you (that is your code after all ;))
But I'll have a look at the first issue, curious where that id comes from...

**Comment by chimcen** [comment]
I'm not shure if no. 2 is a code problem...

**Comment by M4LuZ** [comment]
Nr. 1: That I cannot reproduce. Whatever I save as Server ID is directly referred on the next page refresh
Nr. 2: Fixed. There was actually no need to use the guild data, server returns the same JSON data on both calls.
Nr. 3: Image size is defined on your page in dark-theme.min.css lines 1509 and following. As I just got your code up and running in LS, that of course is not included. If you also want to contribute that, then we should be able to extract the discord-specific bits and also include that as css for the box

**Comment by M4LuZ** [comment]
@chimcen I would need your feedback on points 2+3) to continue here.

**Comment by chimcen** [comment]
Nr. 1 - still remains... could it be a problem because i'm still use lansuite 4.2 with the added code line from [https://github.com/lansuite/lansuite/issues/184#issuecomment-385340123](url) ?

Nr. 2 - works now - thank you :+1: 

Nr. 3 - changed the style.css in my lansuite page allready to fit my own design...

**Comment by M4LuZ** [comment]
I'll give it a try on LS 4.2 myself next week.
Would still be happy if you could confirm that we may use your existing CSS formating

**Comment by chimcen** [comment]
Yes, you could use it - just use this one. It contains the changes to match lansuit theme.
Only the colors should be adjusted a little bit more...


[style.css.txt](https://github.com/lansuite/lansuite/files/2048187/style.css.txt)

**Comment by M4LuZ** [comment]
I tried to reproduce your error in my own 4.2 installation but it took over every modification of the server ID as expected. You sure that you saved the new value?

I included your style.css as default, while allowing a separate definiton by placing a `discord.css` into your design folder.

I opened a merge request for the finished code and would close this issue with that unless somebody complains.

BTW: It seems that there was a guide how to include CSS elements.
But that was in the old documentation, which is no longer available due to #239 
Thus another half an hour of reverse engineering, yay!
